132 CRAN pre-release cleanup

m7pr opened the pull request:
Closes #132 

Hey @donyunardi I know this is not meant to be run this week, however taking into consideration the timing of the availability of the maintainer of the package next week, I though we could give this release a go, especially since the ADS conference is coming up soon.

In this PR I took a stab on cleaning the package and on extending README with installation and usage instructions. Besides that I also:

- added CRAN badges
- deleted dates from vignettes
- fixed github.io references so that we use `/latest-tag/` version of our website
- improved phrasing in narration within vignettes
- run R CMD CHECK
- edited NEWS
- improved README with installation instructions, getting help information and example usage

pawelru commented on README.md:
````diff
@@ -32,18 +38,67 @@ include:
+```r
+library(teal.code)
+my_qenv <- new_qenv(env = list2env(list(x = 5)), code = "x <- 5")
+my_qenv
+Parent: <environment: package:teal.code>
+Bindings:
+• x: <dbl> [L]
+```
+
+```r
+qenv_2 <- eval_code(my_qenv, "y <- x * 2") |> eval_code("z <- y * 2")
+qenv_2
+<environment: 0x00000135b544cfe8> [L]
+Parent: <environment: package:teal.code>
+Bindings:
+• x: <dbl> [L]
+• y: <dbl> [L]
+• z: <dbl> [L]
+```
+
+```r
+qenv_2[["y"]]
+[1] 10
+```
 
 ```r
-Sys.setenv(GITHUB_PAT = "your_access_token_here")
-if (!require("remotes")) install.packages("remotes")
-remotes::install_github("insightsengineering/teal.code@*release")
+cat(paste(get_code(qenv_2), collapse = "\n"))
+x <- 5
+y <- x * 2
+z <- y * 2
 ```
````
would it be possible to somehow distinguish command from the output? Right not it's not super readable:
<img width="320" alt="image" src="https://github.com/insightsengineering/teal.code/assets/12943682/a2bd14e7-c569-410f-b527-cad9f3c6647d">

e.g. `dplyr` does `#>` upfront any of the line in the output
there might be other ways as well

(please pass this on to the other readme updates PRs if exist)

pawelru commented:
Can you also fix one link please? (the last one from the below)
```
r$> urlchecker::url_check()
✖ Error: README.md:5:79 404: Not Found
[![CRAN Version](https://www.r-pkg.org/badges/version/teal.code?color=green)](https://cran.r-project.org/package=teal.code)
                                                                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
✖ Error: README.md:6:90 404: Not Found
[![Total Downloads](http://cranlogs.r-pkg.org/badges/grand-total/teal.code?color=green)](https://cran.r-project.org/package=teal.code)
                                                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
✖ Error: README.md:7:94 404: Not Found
[![Last Month Downloads](http://cranlogs.r-pkg.org/badges/last-month/teal.code?color=green)](https://cran.r-project.org/package=teal.code)
                                                                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
✖ Error: README.md:8:92 404: Not Found
[![Last Week Downloads](http://cranlogs.r-pkg.org/badges/last-week/teal.code?color=green)](https://cran.r-project.org/package=teal.code)
                                                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
! Warning: README.md:37:36 Moved
`teal.code` also ships a [`shiny`](https://shiny.rstudio.com/) module that helps inspect the stored code as well as messages, warnings and error messages resulting from evaluation via `shiny` web application.
                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~
                                   https://shiny.posit.co/
```

pawelru commented:
Moreover, please add `URL` and `BugReports` fields to the DESCRIPTION file

```
r$> goodpractice::gp()
(...)
── GP teal.code ─────────────────────────────────────────────────────────────────

It is good practice to

  ✖ write unit tests for all functions, and all package code in
    general. 95% of code lines are covered by test cases.

    R/include_css_js.R:12:NA
    R/include_css_js.R:13:NA
    R/include_css_js.R:14:NA
    R/include_css_js.R:15:NA
    R/include_css_js.R:16:NA
    ... and 5 more lines

  ✖ omit "Date" in DESCRIPTION. It is not required and it gets
    invalid quite often. A build date will be added to the package when
    you perform `R CMD build` on it.
  ✖ add a "URL" field to DESCRIPTION. It helps users find
    information about your package online. If your package does not
    have a homepage, add an URL to GitHub, or the CRAN package package
    page.
  ✖ add a "BugReports" field to DESCRIPTION, and point it to a
    bug tracker. Many online code hosting services provide bug trackers
    for free, https://github.com, https://gitlab.com, etc.
  ✖ avoid long code lines, it is bad for readability. Also,
    many people prefer editor windows that are about 80 characters
    wide. Try make your lines shorter than 80 characters

    R/qenv-class.R:7:81
    R/qenv-class.R:9:81
    R/qenv-class.R:16:81
    R/qenv-class.R:18:81
    R/qenv-concat.R:3:81
    ... and 116 more lines

  ✖ avoid the attach() and detach() functions, they are fragile
    and code that uses them will probably break sooner than later.

    tests/testthat/test-qenv_eval_code.R:35:3

  ✖ not import packages as a whole, as this can cause name
    clashes between the imported packages. Instead, import only the
    specific functions you need.
───────────────────────────────────────────────────────────────────────────────── 
```

pawelru commented:
Yet another finding - some functionality does not have documented return value
(we should look into auto-test it somehow)

```
r$> x <- pkgcheck::pkgcheck()
r$> summary(x)

── teal.code 0.4.0.9005 ─────────────────────────────────────────────────────────

✔ Package name is available
✖ does not have a 'codemeta.json' file.
✔ has a 'contributing' file.
✖ The following functions have no documented return values: [concat, dev_suppress, eval_code, get_var, join, new_qenv]
✔ uses 'roxygen2'.
✖ 'DESCRIPTION' does not have a URL field.
✖ 'DESCRIPTION' does not have a BugReports field.
✔ Package has at least one HTML vignette
✔ All functions have examples.
✖ Package uses global assignment operator ('<<-').
✖ Continuous integration checks unavailable (no URL in 'DESCRIPTION').
✔ Package coverage is 95.1%.
✔ R CMD check found no errors.
✔ R CMD check found no warnings.
ℹ Function names are duplicated in other packages

ℹ Current status:
✖ This package is not ready to be submitted.
```

pawelru commented:
If you are hardcore player 😈 

```
r$> x <- lintr::lint_package(linters = lintr::all_linters())
........................

r$> length(x)
[1] 322
```

pawelru commented:
> @pawelru (tagging @KlaudiaBB ) should we add
> 
> `urlchecker::url_check()`, `goodpractice::gp()` and `pkgcheck::pkgcheck()` to CRAN release GitHub Issue template?
> 
> Thanks for pointing those @pawelru

I have already requested for `urlchecker` - it's on the IDR tasks.

For the rest - let's think more. These have a lot of false positives and should be treated as _guidelines_ as opposed to a _test_ that fails the workflows. On the other hand there are some useful things in there that we can try to incorporate - not yet sure if via these packages or some other way

m7pr commented:
Hey @pawelru I updated Shiny website address and added URL and BugReports. Thanks for adding a note on it to https://github.com/insightsengineering/coredev-tasks/issues/462 . I also addressed the feedback regarding `pkgcheck`.

@donyunardi you are good to merge and good to start a proper CRAN release issue.

pawelru reviewed (approved):
LGTM 👍
